npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+	c.npMgr.Lock()
+	cachedNetPolObj, netPolExists := c.npMgr.RawNpMap[netPolCachedKey]
+	c.npMgr.Unlock()
+	if netPolExists {
+		// if network policy does not have different states against lastly applied states stored in cachedNetPolObj,
+		// netPolController does not need to reconcile this update.
+		// in this updateNetworkPolicy event, newNetPol was updated with states which netPolController does not need to reconcile.
+		if isSameNetworkPolicy(cachedNetPolObj, newNetPol) {
+			return
+		}
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) deleteNetworkPolicy(obj interface{}) {
+	_, ok := obj.(*networkingv1.NetworkPolicy)
+	// DeleteFunc gets the final state of the resource (if it is known).
+	// Otherwise, it gets an object of type DeletedFinalStateUnknown.
+	// This can happen if the watch is closed and misses the delete event and
+	// the controller doesn't notice the deletion until the subsequent re-list
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type: %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+
+		if _, ok = tombstone.Obj.(*networkingv1.NetworkPolicy); !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type (error decoding object tombstone, invalid type): %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+	}
+
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+
+	// (TODO): need to decouple this lock from npMgr if possible
+	c.npMgr.Lock()
+	_, netPolExists := c.npMgr.RawNpMap[netPolCachedKey]
+	c.npMgr.Unlock()
+	// If a network policy object is not in the RawNpMap, do not need to clean-up states for the network policy
+	// since netPolController did not apply for any states for the network policy
+	if !netPolExists {
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	klog.Infof("Starting Network Policy %d worker(s)", threadiness)
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.Infof("Started Network Policy %d worker(s)", threadiness)
+	<-stopCh
+	klog.Info("Shutting down Network Policy workers")
+
+	return nil
+}
+
+func (c *networkPolicyController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *networkPolicyController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
```

Comment:
>defer c.workqueue.Done(obj) [](start = 2, length = 27)

                   c.workqueue.Done(obj)
                   if err != nil {
                      c.workqueue.AddRateLimited(key)
                  }

Reply:
I think we agree to stick to current code.